src/storage/src/hummock/sstable/multi_builder.rs

```diff
@@ -171,123 +178,142 @@ impl<F: TableBuilderFactory> CapacitySplitTableBuilder<F> {
                 .observe(meta.encoded_size() as _);
 
             let sst_info = SstableInfo {
-                id: sst_id,
+                id: builder_output.sstable_id,
                 key_range: Some(risingwave_pb::hummock::KeyRange {
                     left: meta.smallest_key.clone(),
                     right: meta.largest_key.clone(),
                     inf: false,
                 }),
                 file_size: meta.estimated_size as u64,
-                table_ids,
+                table_ids: builder_output.table_ids,
             };
-            let policy = self.policy;
-            let mut tracker = self.tracker.take().unwrap();
-            let upload_join_handle = tokio::spawn(async move {
-                if !tracker.try_increase_memory(data.capacity() as u64 + meta.encoded_size() as u64)
-                {
-                    tracing::debug!("failed to allocate increase memory for meta file, sst id: {}, file size: {}, meta size: {}",
-                        sst_id, data.capacity(), meta.encoded_size());
-                }
-                let ret = sstable_store.put_sst(sst_id, meta, data, policy).await;
-                drop(tracker);
-                ret
-            });
-            self.sealed_builders.push(SealedSstableBuilder {
+
+            debug_assert!(self.tracker.is_some());
+            let writer_output = builder_output.writer.finish(meta, self.tracker.take())?;
```

Comment:
I think we can move the logic of memory tracking into the `impl` of trait `SstableWriter`, that is, we have a `Arc<MemoryLimiter>` in some implementation of trait `SstableWriterBuilder`, and when it builds a new `SstableWriter`, it can pass a `MemoryTracker` into the `SstableWriter` it generates if necessary. In this way, we don't even need a `get_memory_usage` method in `SstableWriter` trait. cc @Little-Wallace

Reply:
`get_memory_usage` is on `SstableStreamingUploader`, because the memory usage depends on the implementation of object store.

Reply:
I see. So the `get_memory_usage` can be kept in the `StreamingUploader` trait. 

But I think the `SstableStreamingUploader ` does not need to expose a `get_memory_usage` method. Instead, it can accept a `MemoryTracker` and call `tracker.try_increase_memory` by itself when necessary.

Reply:
Then we need to pass the tracker all the way down to obejct store.

Reply:
No. The object store or the streaming uploader returned from object store does not need to hold the memory tracker. The `SstableStreamingUploader` or the `SstableStreamingWriter` will hold the memory tracker, and get its memory usage by collecting its own stats or calling `get_memory_usage` of `BoxedStreamingUploader`, and then call `try_increase_memory` of the memory tracker. 

In this way, the `CapacitySplitTableBuilder` will not be aware of any concept of memory limitation and does not have to hold any memory limiter or tracker. The memory limitation will be controlled via the `SstableWriter` itself when necessary. 

The `impl` of `SstableWriterBuilder` that generates a new `SstableStreamingWriter` can hold an `Arc<MemoryLimiter>`, and then creates a new memory tracker from the limiter and pass the tracker to the `SstableStreamingWriter` when creating it. The tracker will not be passed long way down to somewhere deep.